lib/phoenix_live_view/async_result.ex

```diff
@@ -0,0 +1,501 @@
+         |> assign_async([:profile, :rank], fn -> {:ok, %{profile: ..., rank: ...}} end)}
+      end
+
+  Here we are assigning `:org` and `[:profile, :rank]` asynchronously. The async function
+  must return a `{:ok, assigns}` or `{:error, reason}` tuple, where `assigns` is a map of
+  the keys passed to `assign_async`. If the function returns other keys or a different
+  set of keys, an error is raised.
+
+  The state of the async operation is stored in the socket assigns within an
+  `%AsyncResult{}`. It carries the loading and error states, as well as the result.
+  For example, if we wanted to show the loading states in the UI for the `:org`,
+  our template could conditionally render the states:
+
+  ```heex
+  <div :if={@org.state == :loading}>Loading organization...</div>
+  <div :if={org = @org.ok? && @org.result}}><%= org.name %> loaded!</div>
+  ```
+
+  The `with_state` function component can also be used to declaratively
+  render the different states using slots:
+
+  ```heex
+  <AsyncResult.with_state :let={org} assign={@org}>
+    <:loading>Loading organization...</:loading>
+    <:empty>You don't have an organization yet</:error>
+    <:error :let={{_kind, _reason}}>there was an error loading the organization</:error>
+    <:canceled :let={_reason}>loading canceled</:canceled>
+    <%= org.name %>
+  <AsyncResult.with_state>
+  ```
+
+  Additionally, for async assigns which result in a list of items, you
+  can consume the assign directly. It will only enumerate
+  the results once the results are loaded. For example:
+
+  ```heex
+  <div :for={orgs <- @orgs}><%= org.name %></div>
+  ```
+
+  ## Arbitrary async operations
+
+  Sometimes you need lower level control of asynchronous operations, while
+  still receiving process isolation and error handling. For this, you can use
+  `start_async/3` and the `AsyncResult` module directly:
+
+      def mount(%{"id" => id}, _, socket) do
+        {:ok,
+        socket
+        |> assign(:org, AsyncResult.new(:org))
+        |> start_async(:my_task, fn -> fetch_org!(id) end)
+      end
+
+      def handle_async(:org, {:ok, fetched_org}, socket) do
+        %{org: org} = socket.assigns
+        {:noreply, assign(socket, :org, AsyncResult.ok(org, fetched_org))}
+      end
+
+      def handle_async(:org, {:exit, reason}, socket) do
+        %{org: org} = socket.assigns
+        {:noreply, assign(socket, :org, AsyncResult.exit(org, reason))}
+      end
+
+  `start_async/3` is used to fetch the organization asynchronously. The
+  `handle_async/3` callback is called when the task completes or exists,
+  with the results wrapped in either `{:ok, result}` or `{:exit, reason}`.
+  The `AsyncResult` module is used to direclty to update the state of the
+  async operation, but you can also assign any value directly to the socket
+  if you want to handle the state yourself.
+  '''
+  use Phoenix.Component
+
+  defstruct name: nil,
+            keys: [],
```

Comment:
Those keys should be not part of the struct IMO.

```suggestion
  defstruct name: nil,
            keys: [],
```